### PR TITLE
Move field name validation logic out of mongo

### PR DESCRIPTION
**Pull request by drew-gross**

### CODE REVIEW

File: src/Adapters/Storage/Mongo/MongoTransform.js [hunk 1]
```diff
@@ -71,9 +71,6 @@ function transformKeyValue(schema, className, restKey, restValue, {
     if (authDataMatch) {
       throw new Parse.Error(Parse.Error.INVALID_KEY_NAME, 'can only query on ' + key);
     }
-    if (validate && !key.match(/^[a-zA-Z][a-zA-Z0-9_\.]*$/)) {
```
Comment (flovilmart) [inline src/Adapters/Storage/Mongo/MongoTransform.js]: we don't add the validation anywhere else or was it a duplicate?

File: src/Controllers/SchemaController.js [hunk 2]
```diff
@@ -466,15 +466,16 @@ class SchemaController {
...
-      this._collection.transform.transformKey(this, className, fieldName);
-
-      if( fieldName.indexOf(".") > 0 ) {
+      if (fieldName.indexOf(".") > 0) {
         // subdocument key (x.y) => ok if x is of type 'object'
         fieldName = fieldName.split(".")[ 0 ];
         type = 'Object';
       }
 
+      if (!fieldNameIsValid(fieldName)) {
```
Comment (drew-gross) [inline src/Controllers/SchemaController.js]: Here

File: src/Controllers/DatabaseController.js [hunk 3]
```diff
@@ -618,9 +618,23 @@ DatabaseController.prototype.find = function(className, query, {
...
+        if (fieldName === '_created_at') {
+          fieldName = 'createdAt';
+          sort['createdAt'] = sort['_created_at'];
+        }
+        if (fieldName === '_updated_at') {
+          fieldName = 'updatedAt';
+          sort['updatedAt'] = sort['_updated_at'];
+        }
+
+        if (!SchemaController.fieldNameIsValid(fieldName)) {
```
Comment (drew-gross) [inline src/Controllers/DatabaseController.js]: And here

File: src/Adapters/Storage/Mongo/MongoTransform.js (hunk 1 above)
Comment (drew-gross) [inline src/Adapters/Storage/Mongo/MongoTransform.js]: It was duplicate. Replaced with `fieldNameIsValid`

File: src/Controllers/DatabaseController.js [hunk 4]
```diff
@@ -618,9 +618,23 @@ DatabaseController.prototype.find = function(className, query, {
...
-        let mongoKey = this.transform.transformKey(schemaController, className, key);
-        mongoOptions.sort[mongoKey] = sort[key];
+      for (let fieldName in sort) {
+        // Parse.com treats queries on _created_at and _updated_at as if they were queries on createdAt and updatedAt,
+        // so duplicate that behaviour here.
+        if (fieldName === '_created_at') {
+          fieldName = 'createdAt';
+          sort['createdAt'] = sort['_created_at'];
+        }
+        if (fieldName === '_updated_at') {
```
Comment (peterdotjs) [inline src/Controllers/DatabaseController.js]: could be `else if`

File: src/Controllers/DatabaseController.js [hunk 5]
```diff
@@ -618,9 +618,23 @@ DatabaseController.prototype.find = function(className, query, {
...
+          fieldName = 'updatedAt';
+          sort['updatedAt'] = sort['_updated_at'];
+        }
+
+        if (!SchemaController.fieldNameIsValid(fieldName)) {
+          throw new Parse.Error(Parse.Error.INVALID_KEY_NAME, `Invalid field name: ${fieldName}.`);
+        }
+        let mongoKey = this.transform.transformKeyValue(schemaController, className, fieldName, null).key;
```
Comment (peterdotjs) [inline src/Controllers/DatabaseController.js]: `const`

**Comment by peterdotjs** [comment]
some nits but lgtm.

**Comment by facebook-github-bot-9** [comment]
@drew-gross updated the pull request.

File: src/Adapters/Storage/Mongo/MongoTransform.js (hunk 1 above)
Comment (flovilmart) [inline src/Adapters/Storage/Mongo/MongoTransform.js]: :+1:

**Pull request merged**